modules/rostests/apitests/acpi/Bus_PDO_EvalMethod.c

```diff
@@ -0,0 +1,1547 @@
+/*
+ * PROJECT:     ReactOS API Tests
+ * LICENSE:     LGPL-2.1-or-later (https://spdx.org/licenses/LGPL-2.1-or-later)
+ * PURPOSE:     Unit Tests for acpi!Bus_PDO_EvalMethod (IOCTL_ACPI_EVAL_METHOD handler)
+ * COPYRIGHT:   Copyright 2024 Dmitry Borisov (di.sean@protonmail.com)
+ */
+
+/* INCLUDES *******************************************************************/
+
+#include <apitest.h>
+
+#define WIN32_NO_STATUS
+#include <ndk/rtlfuncs.h>
+
+#define UNIT_TEST
+#include <acpi.h>
+
+#include <acpiioct.h>
+
+#include <initguid.h>
+
+/* TEST DEFINITIONS ***********************************************************/
+
+typedef struct _EVAL_TEST_ENTRY
+{
+    ULONG Line;
+
+    ULONG Flags;
+#define STM_TEST_FLAG_INVALID_ARG_3_1              (1 << 0 )
+#define STM_TEST_FLAG_INVALID_ARG_3_2              (1 << 1 )
+#define STM_TEST_FLAG_INVALID_ARG_3_4              (1 << 2 )
+#define STM_TEST_FLAG_INVALID_ARG_3_5              (1 << 3 )
+#define STM_TEST_FLAG_INVALID_SIZE_1               (1 << 4 )
+#define STM_TEST_FLAG_LARGE_ARG_BUFFER             (1 << 5 )
+#define STM_TEST_FLAG_CHANGE_ARG_COUNT             (1 << 6 )
+#define STM_TEST_FLAG_SUB_IN_BUFFER                (1 << 7 )
+#define STM_TEST_FLAG_SUB_IRP_BUFFER               (1 << 8 )
+#define STM_TEST_FLAG_SET_IN_BUFFER                (1 << 9 )
+#define STM_TEST_FLAG_SET_IRP_BUFFER               (1 << 10)
+#define STM_TEST_FLAG_BAD_ARG_TYPE                 (1 << 11)
+#define STM_TEST_FLAG_IGNORE_FAILURE               (1 << 12)
+
+#define GTM_TEST_FLAG_BAD_SIGNARUTE                (1 << 0)
+#define GTM_TEST_FLAG_BUFFER_HAS_SIGNARUTE         (1 << 1)
+#define GTM_TEST_FLAG_BUFFER_HAS_COUNT             (1 << 2)
+#define GTM_TEST_FLAG_BUFFER_HAS_LENGTH            (1 << 3)
+#define GTM_TEST_FLAG_ARG_HAS_BUFFER_TYPE          (1 << 4)
+#define GTM_TEST_FLAG_ARG_HAS_DATA_LENGTH          (1 << 5)
+#define GTM_TEST_FLAG_INC_OUT_BUFFER               (1 << 6)
+#define GTM_TEST_FLAG_DEC_OUT_BUFFER               (1 << 7)
+#define GTM_TEST_FLAG_SET_OUT_BUFFER               (1 << 8)
+
+#define GTM_TEST_FLAG_METHOD_SUCCESS \
+    (GTM_TEST_FLAG_BUFFER_HAS_SIGNARUTE | \
+     GTM_TEST_FLAG_BUFFER_HAS_COUNT | GTM_TEST_FLAG_BUFFER_HAS_LENGTH | \
+     GTM_TEST_FLAG_ARG_HAS_BUFFER_TYPE | GTM_TEST_FLAG_ARG_HAS_DATA_LENGTH)
+
+#define DSM_TEST_FLAG_EMPTY_PACKAGE                (1 << 0)
+#define DSM_TEST_FLAG_LARGE_SUB_PACKAGE_BUFFER     (1 << 1)
+
+    NTSTATUS Status;
+    ULONG Value;
+} EVAL_TEST_ENTRY;
+
+/* KERNEL DEFINITIONS (MOCK) **************************************************/
+
+#define PAGED_CODE()
```

Comment:
I think it may be worth putting these Kernel definitions/minimal reimplementation in user-mode into a separate file, for reusability purposes.

Reply:
This change should preferably be done in another PR.

Reply:
ok